File: MvpFramework/src/main/java/com/ytx/mvpframework/presenter/BasePresenter.java
package com.ytx.mvpframework.presenter;

import com.ytx.mvpframework.view.IView;

import rx.Subscription;
import rx.subscriptions.CompositeSubscription;

public class BasePresenter<V extends IView> {
    private CompositeSubscription compositeSubscription = null;
    private V view;

    public BasePresenter(V view) {
        this.view = view;
    }

    public final V getView() {
        return view;
    }

    public final void addSubscription(Subscription subscription) {
        if (compositeSubscription == null || compositeSubscription.isUnsubscribed()) {
            synchronized(BasePresenter.class) {
                if (compositeSubscription == null || compositeSubscription.isUnsubscribed()) {
                    compositeSubscription = new CompositeSubscription();
                }
            }
        }
        compositeSubscription.add(subscription);
    }

    public final void unSubscribe() {
        if (compositeSubscription != null) {
            compositeSubscription.unsubscribe();
        }
    }
}

File: MvpFramework/src/main/java/com/ytx/mvpframework/presenter/BasePresenter.kt
//@file:JvmName("BasePresenter")
//package com.ytx.mvpframework.presenter
//
//import android.arch.lifecycle.LifecycleObserver
//import com.ytx.mvpframework.view.IView
//import rx.Subscription
//import rx.subscriptions.CompositeSubscription
//
//abstract class BasePresenter<V : IView>(val view: V) : LifecycleObserver {
//    private var compositeSubscription: CompositeSubscription? = null
//
//    fun addSubscription(subscription: Subscription) {
//        if (compositeSubscription == null || compositeSubscription!!.isUnsubscribed) {
//            synchronized(1) {
//                if (compositeSubscription == null || compositeSubscription!!.isUnsubscribed) {
//                    compositeSubscription = CompositeSubscription()
//                }
//            }
//        }
//        compositeSubscription!!.add(subscription)
//    }
//
//    fun unSubscribe() {
//        compositeSubscription?.unsubscribe()
//    }
//}